### PR TITLE
Initial separation of dparray into 2 parts.

### PR DESCRIPTION
**Pull request by DrTodd13**
This isn't complete since I wasn't able to get the file showing up as part of the package so I wasn't able to test much.  Help needed!

### CODE REVIEW

**Comment by PokhodenkoSA** [comment]
Need tests. @vlad-perevezentsev could write tests. @DrTodd13 could you provide an idea how to test this?
Possibly dpNP had tests for dparray.

**Comment by PokhodenkoSA** [comment]
This PR have an issue to fix: can not import dparray from dpctl.

**Comment by PokhodenkoSA** [comment]
Some tests for dparray could be taken from https://github.com/IntelPython/numba/pull/116/files#diff-74090bd7024ff78832de33f87c13b014c1646982a2c286e6200913699be5d91e.

**Comment by PokhodenkoSA** [comment]
Related to #188

**Comment by oleksandr-pavlyk** [comment]
What's the best way to push changes to this branch?

File: dpctl/dparray.py [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+
+debug = False
+
+def dprint(*args):
+    if debug:
+        print(*args)
+        sys.stdout.flush()
+
+import dpctl
+from dpctl._memory import MemoryUSMShared
```
Comment (oleksandr-pavlyk) [inline dpctl/dparray.py:18]: Should be `from dpctl.memory import MemoryUSMShared`.

File: dpctl/dparray.py [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+
+functions_list = [o[0] for o in getmembers(np) if isfunction(o[1]) or isbuiltin(o[1])]
+class_list = [o for o in getmembers(np) if isclass(o[1])]
+
+array_interface_property = "__array_interface__"
```
Comment (oleksandr-pavlyk) [inline dpctl/dparray.py:23]: Should be `__sycl_usm_array_interface__` instead.

File: dpctl/dparray.py [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+    def __new__(subtype, shape,
+                dtype=float, buffer=None, offset=0,
+                strides=None, order=None):
+        # Create a new array.
+        if buffer is None:
+            dprint("dparray::ndarray __new__ buffer None")
+            nelems = np.prod(shape)
+            dt = np.dtype(dtype)
+            isz = dt.itemsize
+            buf = MemoryUSMShared(nbytes=isz*max(1,nelems))
```
Comment (oleksandr-pavlyk) [inline dpctl/dparray.py:41]: should be 

```python
nbytes = int(isz*max(1, nelems))
buf = MemoryUSMShared(nbytes)
```

File: dpctl/dparray.py [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+            return new_obj
+        else:
+            dprint("dparray::ndarray __new__ buffer not None and not sycl_usm")
+            nelems = np.prod(shape)
+            # must copy
+            ar = np.ndarray(shape,
+                            dtype=dtype, buffer=buffer,
+                            offset=offset, strides=strides,
+                            order=order)
+            buf = MemoryUSMShared(nbytes=ar.nbytes)
```
Comment (oleksandr-pavlyk) [inline dpctl/dparray.py:74]: Same as above, should be 

```python
nbytes = int(ar.nbytes)
buf = MemoryUSMShared(nbytes)
```

File: dpctl/dparray.py [hunk 5]
```diff
@@ -0,0 +1,222 @@
...
+        dprint("__array_finalize__:", obj, hex(id(obj)), type(obj))
+        # When called from the explicit constructor, obj is None
+        if obj is None: return
+        # When called in new-from-template, `obj` is another instance of our own
+        # subclass, that we might use to update the new `self` instance.
+        # However, when called from view casting, `obj` can be an instance of any
+        # subclass of ndarray, including our own.
+        if hasattr(obj, array_interface_property):
+            return
+        if isinstance(obj, numba.core.runtime._nrt_python._MemInfo):
```
Comment (oleksandr-pavlyk) [inline dpctl/dparray.py:97]: `numba` is undefined.

**Comment by oleksandr-pavlyk** [comment]
Method `__call__` is handled, in `__array_ufunc__`, but method ``__reduce__`` is not:

```
In [7]: ms = dpctl.memory.MemoryUSMShared(8*4)

In [8]: X = dpctl.dparray.ndarray((4,), dtype='d', buffer=ms)

In [9]: X
Out[9]:
ndarray([4.64247707e-310, 4.64247698e-310, 9.48606040e-322,
         2.37151510e-322])

In [10]: np.add(X, 1)
Out[10]: ndarray([1., 1., 1., 1.])

In [11]: np.add(X, 1).base
Out[11]: <Intel(R) USM allocated memory block of 32 bytes at 0x5575e19ea000>

In [12]: X.sum()
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-12-16f51ebe4109> in <module>()
----> 1 X.sum()

/home.conda/envs/idp/lib/python3.7/site-packages/numpy/core/_methods.py in _sum(a, axis, dtype, out, keepdims, initial, where)
     36 def _sum(a, axis=None, dtype=None, out=None, keepdims=False,
     37          initial=_NoValue, where=True):
---> 38     return umr_sum(a, axis, dtype, out, keepdims, initial, where)
     39
     40 def _prod(a, axis=None, dtype=None, out=None, keepdims=False,

TypeError: operand type(s) all returned NotImplemented from __array_ufunc__(<ufunc 'add'>, 'reduce', ndarray([4.64247707e-310, 4.64247698e-310, 9.48606040e-322,
         2.37151510e-322]), axis=None, dtype=None, keepdims=False, where=True): 'ndarray'

```

**Comment by PokhodenkoSA** [comment]
@oleksandr-pavlyk I will fix all your review and notes tomorrow. This PR is in the core of many further modifications and I would like to merge it tomorrow after fixing and tests adding.

File: dpctl/dparray.py (hunk 5 above)
Comment (PokhodenkoSA) [reply dpctl/dparray.py:97]: @DrTodd13 Why we use this function from numba? Is it optional? Can we copy this function to our code?

**Comment by PokhodenkoSA** [comment]
> Method `__call__` is handled in `__array_ufunc__`, but method `__reduce__` is not:

@oleksandr-pavlyk  @DrTodd13 How it could be fixed? Is it ok to create a separate issue for it and fix in separate PR?

**Pull request merged**